### DISCUSSION TITLE
Git repo for forms not quite working.

### DISCUSSION THREAD

**Discussion by slekkus75** [ansibleguy76/ansibleforms#179] [Q&A] [answered]
Changed forms to a git repo and added this repo, all fine. 

I have all my forms and forms.yaml in gitlab repo and this syncs to below folder fine.
"/app/dist/persistent/repositories/mygitrepo_folder_forms" where I do have some forms I use. 

But refresh (as admin) doesn't show anything. blank screen. If I do move forms.yaml out of this folder, I get to see my forms.
Question is, where does forms.yaml need to be in?

Couldn't find any info in the documentation on how this should work .

regards..

**Comment by ansibleguy76** [comment]
the forms.yaml should be in the original place (non-repo).  forms.yaml should then only contain the roles and categories.  Or that's the idea, to move the actual "forms" to repo's and keep the forms.yaml internal.  It you can think of a better way, please share and I'll concider it.  But note that I need to be able to process the roles even if all repo's are failing.  It's a bit of legacy really.  originally it was only 1 forms files, then I added the forms folder and then the repos.  I could potentially move the roles and categories to the database for example or to a "config" file.  But I want to be as backwards compatible as possible

**Reply by ansibleguy76** [reply]
so does this require any further followup?

**Comment by slekkus75** [comment]
Good day, sorry for the delay. I understand now where this originated from. 
Not sure I am the person to propose ideas (barely able to get this working), but placing the roles in the database and thus configurable from the settings menu, would be logical and more practical. Categories tend not to change that often.

btw, in the log it seems, forms.yaml is searched for in both locations. 
```
2024-06-14 14:55:16:5516 info: [ansibleforms] running query : SELECT name FROM AnsibleForms.`repositories` WHERE use_for_forms
22024-06-14 14:55:16:5516 error: Failed to get forms path :  ENOENT: no such file or directory, access '/app/dist/persistent/repositories/mygitrepo_folder_forms/forms.yaml'
2024-06-14 14:55:16:5516 info: Loading /app/dist/persistent/forms.yaml
```

**Reply by ansibleguy76** [reply]
i could do that, and I thought about it, but you need to be admin to get to the settings.  If you lock yourself out, you can't correct it.
Having it in the yaml file, allows you to change it outside the database.

But there would be ways to work around that.  Move the roles to the databsase and allow to bypass the login process with environment variables in case of emergency.  That would work.  Same for categories.

This way, you can login without having a single valid forms file, setting up a repo and pull.

**Reply by slekkus75** [reply]
I see. Would an environment variable, which when set to true (e.g. USE_FORMS_CONFIG) to ignore database settings also work?
This way users get to choose how they use it, and as a failsafe for default (database users). For the latter, they'd need to keep an updated forms.yaml. Or always keep both, and every change in DB write changes to forms.yaml too.

**Reply by ansibleguy76** [reply]
yes, absolutely...  we write the code ourselves :) so everything is possible.  I will think about it.  So bottomline, you want the option to have no forms.yaml.  You just have the database config and your repo's.  That's fair and makes sense.  Only the potential admin lock-out needs to be tackled.  But I can build an admin pass-through env variable, that logs you in as admin for worse case scenario's.

**Reply by slekkus75** [reply]
That would be nice.
Thanks for the awesome project, enjoy using it.

**Reply by ansibleguy76** [reply]
> Good day, sorry for the delay. I understand now where this originated from. Not sure I am the person to propose ideas (barely able to get this working), but placing the roles in the database and thus configurable from the settings menu, would be logical and more practical. Categories tend not to change that often.
> 
> btw, in the log it seems, forms.yaml is searched for in both locations.
> 
> ```
> 2024-06-14 14:55:16:5516 info: [ansibleforms] running query : SELECT name FROM AnsibleForms.`repositories` WHERE use_for_forms
> 22024-06-14 14:55:16:5516 error: Failed to get forms path :  ENOENT: no such file or directory, access '/app/dist/persistent/repositories/mygitrepo_folder_forms/forms.yaml'
> 2024-06-14 14:55:16:5516 info: Loading /app/dist/persistent/forms.yaml
> ```

I double checked this, and indeed, if the forms.yaml is found in the repo, it will be taken from there.

**Reply by ansibleguy76** [reply]
I have tested this again (with the old code) and it works fine on my end.

forms.yaml loaded from repo as well as the /forms/ folder with several yaml files

**Comment by ansibleguy76** [comment]
So bottomline :

- add forms.yaml and a forms folder (containing loose yaml files with forms) in your repo
- enable "for forms" on the repo

The designer will become disabled (as precaution to prevent conflicts)

- in the new beta, you will find a "forms yaml" section, which will be overwrite the forms.yaml file in case you rather have the base file in the database)

**Reply by ansibleguy76** [reply, marked as answer]
![image](https://github.com/ansibleguy76/ansibleforms/assets/89778474/6f61a988-c5be-42f9-bd03-a6d0a20e6c93)

So in the settings you can now define the base forms.yaml in the database.  To make it easy, you can load if from the original file if present.  This way, the forms.yaml can be dropped from repo.  Don't forget to place your other form-yaml-files under the forms/ folder in your repository.

**Reply by ansibleguy76** [reply]
This is now being pushed to beta 5.0.3.  
Since placing the forms.yaml in the repo works fine, having it in the database might not be "so interesting" for everyone, hence you need to enable this feature with an ENV VAR = ENABLE_FORMS_YAML_IN_DATABASE=1

**Reply by ansibleguy76** [reply]
In case you lock yourself out, there is now a ENABLE_BYPASS env var that will enable temp admin login.

**Comment by slekkus75** [comment]
Thanks for adding the possibility. Will test this once I have more time. Tried it quick (beta 5.0.3) but can't see the forms yaml section in the settings-AF. Updating this settings prompts a invalid data msg also. No idea. Nothing in the log.

**Reply by ansibleguy76** [reply]
to not confuse other users, i made it a hidden feature. you must enable it with an env variable.  see the help.   let me check that message.  maybe i made the url mandatory.  currently on holiday 😎

**Reply by slekkus75** [reply]
Gotya! missed that env var part.
Enjoy your holiday

**Reply by ansibleguy76** [reply]
every tested this ?  Is this what you wanted ?

**Comment by slekkus75** [comment]
Hope you've had a good vacation. I just tried it and works as you intended. This is handy as one does not have to go through git commands to change a simple role. 
And, yes, you have made that url setting mandatory (not sure what its function is :) but pasted in the local url.

**Reply by ansibleguy76** [reply]
url is for notifications (link to job)

**Comment by slekkus75** [comment]
Solved.
